bridges/snowbridge/parachain/primitives/core/src/outbound.rs

```diff
@@ -0,0 +1,408 @@
+use codec::{Decode, Encode};
+use frame_support::PalletError;
+use scale_info::TypeInfo;
+use sp_arithmetic::traits::{BaseArithmetic, Unsigned};
+use sp_core::{RuntimeDebug, H256};
+pub use v1::{AgentExecuteCommand, Command, Initializer, Message, OperatingMode, QueuedMessage};
+
+/// Enqueued outbound messages need to be versioned to prevent data corruption
+/// or loss after forkless runtime upgrades
+#[derive(Encode, Decode, TypeInfo, Clone, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub enum VersionedQueuedMessage {
+	V1(QueuedMessage),
+}
+
+impl TryFrom<VersionedQueuedMessage> for QueuedMessage {
+	type Error = ();
+	fn try_from(x: VersionedQueuedMessage) -> Result<Self, Self::Error> {
+		use VersionedQueuedMessage::*;
+		match x {
+			V1(x) => Ok(x),
+		}
+	}
+}
+
+impl<T: Into<QueuedMessage>> From<T> for VersionedQueuedMessage {
+	fn from(x: T) -> Self {
+		VersionedQueuedMessage::V1(x.into())
+	}
+}
+
+mod v1 {
+	use crate::{pricing::UD60x18, ChannelId};
+	use codec::{Decode, Encode};
+	use ethabi::Token;
+	use scale_info::TypeInfo;
+	use sp_core::{RuntimeDebug, H160, H256, U256};
+	use sp_std::{borrow::ToOwned, vec, vec::Vec};
+
+	/// A message which can be accepted by implementations of `/[`SendMessage`\]`
+	#[derive(Encode, Decode, TypeInfo, Clone, RuntimeDebug)]
+	#[cfg_attr(feature = "std", derive(PartialEq))]
+	pub struct Message {
+		/// ID for this message. One will be automatically generated if not provided.
+		///
+		/// When this message is created from an XCM message, the ID should be extracted
+		/// from the `SetTopic` instruction.
+		///
+		/// The ID plays no role in bridge consensus, and is purely meant for message tracing.
+		pub id: Option<H256>,
+		/// The message channel ID
+		pub channel_id: ChannelId,
+		/// The stable ID for a receiving gateway contract
+		pub command: Command,
+	}
+
+	#[derive(Copy, Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+	pub enum OperatingMode {
+		Normal,
+		RejectingOutboundMessages,
+	}
```

Comment:
nit: please document that this is for operating mode of the beefy light client, and also the concrete purpose of `RejectingOutboutMessages` operation mode (upgrade, maintenance, recovery etc.): something corresponding to https://github.com/paritytech/polkadot-sdk/blob/7e4fe1bea314381f612cb791d3a69f9574cdd24b/bridges/primitives/messages/src/lib.rs#L56-L67 also dumb q: beefy light client handles the polkadot side operation mode being normal (`bp_runtime::BasicOperatingMode::Normal`) or completely halted (`bp_runtime::BasicOperatingMode::Halted`) the same, but `RejectingOutboundMessages` (which, in my understanding, falls somewhere between normal & halted) is handled distinctly?

Reply:
Good point, will improve the documentation for that particular struct.

To clarify, that particular struct is for setting the operating mode of the [Gateway](https://github.com/Snowfork/snowbridge/blob/main/contracts/src/Gateway.sol) contract on Ethereum. The Beefy light client contract, which is a dependency of the Gateway, doesn't have any operating modes, by design.

The Gateway only has two modes  `Normal`, `RejectingOutboundMessages`, for the following reasons:
* Due to the fact that the Gateway is entirely governed by the Polkadot side, we can never allow inbound governance messages to be rejected, otherwise the bridge will essentially be bricked.
* We can also stop the flow of messages to the Gateway from Polkadot by halting the outbound-queue pallet on BridgeHub.

For these reasons, we opted no to include a `Halted` or a `RejectingInboundMessages` mode for the Gateway.